Add validation to licence fields

1pretz1 opened the pull request:
Adds validation rules as specified in the [Trello card](https://trello.com/c/uf3E7dCl/1974-add-validation-to-new-licence-document-fields). 

More info in the commits ->

## Before

<img width="593" alt="Screenshot 2023-06-22 at 18 13 58" src="https://github.com/alphagov/specialist-publisher/assets/24479188/f7fff05e-cbb7-47a8-90e8-c6b21f8f5c2d">

## After

<img width="578" alt="Screenshot 2023-06-26 at 15 22 25" src="https://github.com/alphagov/specialist-publisher/assets/24479188/c53ccdcf-450d-4046-845c-1ee17169084f">
<img width="618" alt="Screenshot 2023-06-26 at 15 16 48" src="https://github.com/alphagov/specialist-publisher/assets/24479188/4b7d901c-97ff-4e51-947a-d099e37952ad">
<img width="621" alt="Screenshot 2023-06-26 at 15 17 45" src="https://github.com/alphagov/specialist-publisher/assets/24479188/0a5e0d83-0560-474b-8b57-d2502fa37ff4">
<img width="618" alt="Screenshot 2023-06-26 at 15 17 59" src="https://github.com/alphagov/specialist-publisher/assets/24479188/66cdffe1-a0ec-47a7-a7f6-6bd7dc69a3f4">
<img width="613" alt="Screenshot 2023-06-26 at 15 19 21" src="https://github.com/alphagov/specialist-publisher/assets/24479188/e309a24a-1362-4e7f-9aa5-cc5f1c900e89">
<img width="620" alt="Screenshot 2023-06-26 at 15 19 45" src="https://github.com/alphagov/specialist-publisher/assets/24479188/c82aafcd-9b79-4770-a821-6d8c29c4fa26">
<img width="598" alt="Screenshot 2023-06-26 at 15 20 14" src="https://github.com/alphagov/specialist-publisher/assets/24479188/5e6f9af6-bec6-45e6-92b6-1b9fe0a33424">


Trello:
https://trello.com/c/uf3E7dCl/1974-add-validation-to-new-licence-document-fields

⚠️ This repo is Continuously Deployed: make sure you [follow the guidance](https://docs.publishing.service.gov.uk/manual/development-pipeline.html#merge-your-own-pull-request) ⚠️

Follow [these steps](https://guides.rubyonrails.org/upgrading_ruby_on_rails.html) if you are doing a Rails upgrade.

1pretz1 committed:
Add LinkOrIdentifierValidator

We expect a licence to have a licence identifier or a link AND will
continue on text.

1pretz1 committed:
Add custom error locales for basic licence fields

Allows a completely custom error message to be used in for the listed
fields for certain validation types. Currently, the implementation is a
bit clumsy but this is a consequence of custom error messages.

We'll add more fields to this list bespoke to licences and there will be
a follow up piece of work to convert all the error messages over to the
new and recommended design system style.

1pretz1 committed:
Adds uniqueness validator for licence identifier

We expect each licence to either have a blank licence identifier (if
linked to an external website) or a unique identifier if managed by a
local authority. The validator pulls in all the licences from Publishing
API and checks whether the identifier, if present, already exists and
fails if it does.

1pretz1 committed:
Adds presence validation for industry / location

Ensures at least one industry and location are selected. Also updates
the validation text to meet the new design pattern for industry /
location and primary publishing organisation.

1pretz1 committed:
Adds validation for continuation link format

Ensures link is valid.

1pretz1 committed:
Use validator in licence transaction model

1pretz1 committed:
Add unique validator to licence transaction model

KludgeKML commented on app/validators/licence_identifier_unique_validator.rb:
```diff
@@ -0,0 +1,15 @@
+
+    existing_licence_identifiers = []
+
+    record.class.find_each do |licence|
```
How fast is this? Is it making 309 api calls every time we try to save a licence?

KludgeKML replied:
(Update - it's more like 8 API calls, as they're done in batches of 50, and there isn't currently a good optimisation we can do here. Appears to be quick in integration, we should just accept this is not perfect but necessary and keep an eye on whether it causes slowdowns in use. Marking as resolved).

KludgeKML reviewed (approved):
LGTM (Pending content approval of warning messages)

1pretz1 committed:
Update error heading text

Advised text: https://design-system.service.gov.uk/components/error-summary/.

1pretz1 committed:
Remove unused partial

1pretz1 committed:
Add help text for inputting licence id/link

Also adds margin top / bottom to field context section which looks
better IMO.

1pretz1 committed:
Update field headings after content review

1pretz1 committed:
Allow custom error messages

We want to use error messages in accordance with the design system
recommendations for the new licence document type. However, adding
custom error messages isn't really supported in Specialist Publisher due
to the default "<attribute> <error_message>" outputted (full_messages).

We want to swap over all the error messages to the new pattern but we're
just implementing this for the licence document type for now.

Also now fetches error messages from locale files which are more
maintainable.

MartinJJones commented on app/views/metadata_fields/_licences.html.erb:
```diff
@@ -28,11 +28,17 @@
 
-<%= render layout: "shared/form_group", locals: { f: f, field: :licence_transaction_will_continue_on, label: "Name of website where you apply" } do %>
+<div class="additional-field-context-section">
+  <h2>How users apply</h3>
```
Small thing, the browser will likely fix this anyway when rendering the page, but the opening tag here is a `h2`, and the closing tag is a `h3`.

1pretz1 replied:
Good spot, now fixed!

MartinJJones reviewed (commented):
Changes look good to me overall, just one small thing to fix 👍

1pretz1 committed:
Show link/identifier errors at the related fields

We now display the link/identifier exists (or blank) errors at the
field group they apply to. Previously we only displayed these errors at
the main error summary at the top of the page.

Also moves the 'How users apply' section to the bottom of page which
makes the grouping more obvious.

KludgeKML commented on app/validators/link_or_identifier_validator.rb:
```diff
@@ -0,0 +1,37 @@
+class LinkOrIdentifierValidator < ActiveModel::Validator
+  def validate(record)
+    if link_and_identifier_exists?(record) || link_and_identifier_missing?(record)
```
Could there be a better name for these functions? Because it reads like: trigger an error if either link and id exists, or link and id are missing.

1pretz1 replied:
Sure, we can rejig the naming if it's not clear, however your assumption of 'trigger an error if either link and id exists, or link and id are missing.'  is correct. We only want to trigger an alert if:

a) the link AND identifier fields are both filled in or 
b) the link AND identifier fields are both blank. 

A user filling in one OR the other is the correct procedure.

How about, `both_link_and_identifier_exist? || both_link_and_identifier_missing?` - thoughts? Happy for other suggestions too!

KludgeKML replied:
Yeah, I like those new names. Nice.

1pretz1 replied:
Done!

MartinJJones reviewed (approved):
Frontend changes look good to me 👍

KludgeKML reviewed (approved):
Looks good to me (with caveat of the variable names)
